workflow/controller/controller.go

```diff
@@ -113,39 +105,107 @@ func (wfc *WorkflowController) Run(ctx context.Context) error {
-	if err != nil {
-		log.Errorf("Failed to register watch for Workflow resource: %v", err)
-		return err
+	// Wait for all involved caches to be synced, before processing items from the queue is started
+	for _, informer := range []cache.SharedIndexInformer{wfc.wfInformer, wfc.podInformer} {
+		if !cache.WaitForCacheSync(ctx.Done(), informer.HasSynced) {
+			return errors.InternalError("Timed out waiting for caches to sync")
+		}
+	}
+
+	wfc.StartStatsTicker(5 * time.Minute)
+	for i := 0; i < 4; i++ {
+		go wait.Until(wfc.runWorker, time.Second, ctx.Done())
+	}
+	for i := 0; i < 8; i++ {
+		go wait.Until(wfc.podWorker, time.Second, ctx.Done())
+	}
+	<-ctx.Done()
+	return ctx.Err()
+}
+
+func (wfc *WorkflowController) runWorker() {
+	for wfc.processNextItem() {
+	}
+}
+
+// processNextItem is the worker logic for handling workflow updates
+func (wfc *WorkflowController) processNextItem() bool {
+	key, quit := wfc.wfQueue.Get()
+	if quit {
+		return false
 	}
+	defer wfc.wfQueue.Done(key)
 
-	// Watch pods related to workflows
-	_, err = wfc.watchWorkflowPods(ctx)
+	obj, exists, err := wfc.wfInformer.GetIndexer().GetByKey(key.(string))
 	if err != nil {
-		log.Errorf("Failed to register watch for Workflow resource: %v", err)
-		return err
+		log.Errorf("Failed to get workflow '%s' from informer index: %+v", key, err)
+		return true
+	}
+	if !exists {
+		// This happens after a workflow was labeled with completed=true
+		// or was deleted, but the work queue still had an entry for it.
+		return true
 	}
+	wf, ok := obj.(*wfv1.Workflow)
+	if !ok {
+		log.Warnf("Key '%s' in index is not a workflow", key)
+		return true
+	}
+	wfc.operateWorkflow(wf)
+	// TODO: operateWorkflow should return error if it was unable to operate properly
+	// so we can requeue the work for a later time
+	// See: https://github.com/kubernetes/client-go/blob/master/examples/workqueue/main.go
+	//c.handleErr(err, key)
+	return true
+}
 
-	i := 0
-	for {
-		if i%100 == 0 {
-			// periodically print the channel sizes
-			i += 1
-			log.Infof("wfChan=%d/%d podChan=%d/%d", len(wfc.wfUpdates), cap(wfc.wfUpdates), len(wfc.podUpdates), cap(wfc.podUpdates))
-		}
-		select {
-		case wf := <-wfc.wfUpdates:
-			wfc.operateWorkflow(wf)
-		case pod := <-wfc.podUpdates:
-			wfc.handlePodUpdate(pod)
-		}
+func (wfc *WorkflowController) podWorker() {
+	for wfc.processNextPodItem() {
 	}
+}
 
-	<-ctx.Done()
-	return ctx.Err()
+// processNextPodItem is the worker logic for handling pod updates.
+// For pods updates, this simply means to up the workflow worker
+// by adding the corresponding entry for the workflow into the
+// workflow workqueue.
+func (wfc *WorkflowController) processNextPodItem() bool {
+	key, quit := wfc.podQueue.Get()
+	if quit {
+		return false
+	}
+	defer wfc.podQueue.Done(key)
+
+	obj, exists, err := wfc.podInformer.GetIndexer().GetByKey(key.(string))
+	if err != nil {
+		log.Errorf("Failed to get pod '%s' from informer index: %+v", key, err)
+		return true
```

Comment:
What does the return value indicate? If it indicates whether the entry from the queue was correctly processed or not, shouldn't this return false?

Reply:
true indicates to the caller we should the control loop. A false means it should stop. This is based off their example and actual controllers

Reply:
IMO, this is a bit unintuitive. What do you think about moving all the code from this method into podWorker itself? The for loop could simply be an infinite loop and instead of return false, there could be a break. This would make it clear that the loop is getting terminated on some events (like quit)

Reply:
Take a look at all the controllers in kubernetes/pkg/controller/*/*_controller.go. This is how kubernetes models all their controllers. I rather follow their lead since it will make it more understandable for people who have written other controllers.

Reply:
Aah, cool!

One of them called namespace_controller (`kubernetes/pkg/controller/namespace/namespace_controller.go`)  does something similar to what was mentioned above:


* Run() creates the workers
```
        for i := 0; i < workers; i++ {
                go wait.Until(nm.worker, time.Second, stopCh)
        }
```

* The nm.worker method has all the required code within it.
```
// worker processes the queue of namespace objects.
// Each namespace can be in the queue at most once.
// The system ensures that no two workers can process
// the same namespace at the same time.
func (nm *NamespaceController) worker() {
        workFunc := func() bool {
                key, quit := nm.queue.Get()
                if quit {
                        return true
                }
                defer nm.queue.Done(key)

                // more logic here
                return false
        }

        for {
                quit := workFunc()

                if quit {
                        return
                }
        }
}
```

Since `workFunc` doesn't have it's own independent existence, it doesn't have any unintuitive behavior. This model works too.